presto-array/src/main/java/com/facebook/presto/array/ReferenceCountMap.java

```diff
@@ -59,19 +64,44 @@ public long sizeOf()
         return INSTANCE_SIZE + SizeOf.sizeOf(key) + SizeOf.sizeOf(value) + SizeOf.sizeOf(used);
     }
 
-    private static final class ObjectStrategy
-            implements Strategy<Object>
+    /**
+     * Get the 64-bit hash code for an object
+     */
+    private static long getHashCode(Object key)
     {
-        @Override
-        public int hashCode(Object object)
-        {
-            return System.identityHashCode(object);
+        // identityHashCode does not guarantee the uniqueness of an object. We need some extra identity.
+        // In the most cases below, we use size of an object to be the extra identity.
+        // Experiments show that with 100 million objects created, using identityHashCode has a collision rate around 2.5%.
+        // However, if these 100 million objects are combined with 10 different sizes, the collision rate is around 0.1%.
+        // The collision rate can be lower than 0.001% if there are 1000 different sizes.
+        int extraIdentity;
+        if (key instanceof Block) {
+            extraIdentity = (int) ((Block) key).getRetainedSizeInBytes();
         }
-
-        @Override
-        public boolean equals(Object left, Object right)
-        {
-            return left == right;
+        else if (key instanceof Slice) {
+            extraIdentity = System.identityHashCode(((Slice) key).getBase());
+        }
+        else if (key instanceof int[]) {
+            extraIdentity = ((int[]) key).length;
```

Comment:
```java
else if (key.getClass().isArray()) {
  extraIdentity = java.lang.reflect.Array.length(value);
}

Reply:
Added benchmark. Using `instanceof` and `reflect` have little difference with each other. Also did the same benchmark against the old code. It is about 50% slower.
```
# Use instanceof int[]/short[]/byte[]/...

Benchmark                                  (arrayType)   Mode  Cnt        Score        Error   Units
BenchmarkReferenceCountMap.benchmarkInserts        int  thrpt   40  5899204.064 ± 296821.134   ops/s
BenchmarkReferenceCountMap.benchmarkInserts     double  thrpt   40  5944614.803 ± 210650.566   ops/s
BenchmarkReferenceCountMap.benchmarkInserts       long  thrpt   40  6031930.357 ± 216640.537   ops/s
BenchmarkReferenceCountMap.benchmarkInserts       byte  thrpt   40  5492756.236 ± 212875.282   ops/s
```

````
# Use reflect.Array

Benchmark                                  (arrayType)   Mode  Cnt        Score        Error   Units
BenchmarkReferenceCountMap.benchmarkInserts        int  thrpt   40  5601146.332 ± 273286.055   ops/s
BenchmarkReferenceCountMap.benchmarkInserts     double  thrpt   40  5595755.182 ± 186303.689   ops/s
BenchmarkReferenceCountMap.benchmarkInserts       long  thrpt   40  5641304.505 ± 195439.326   ops/s
BenchmarkReferenceCountMap.benchmarkInserts       byte  thrpt   40  5772785.212 ± 238556.052   ops/s
```

Reply:
Can you try benchmarking with bulk warmup mode?